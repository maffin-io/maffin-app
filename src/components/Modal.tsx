import React from 'react';

export type ModalProps = {
  open: boolean;
  setOpen: Function;
  title?: string;
  children: React.ReactNode;
};

export default function Modal({
  open, setOpen, title, children,
}: ModalProps): JSX.Element {
  if (!open) {
    return <span />;
  }

  return (
    <div
      className="fixed inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
      id="my-modal"
    >
      <div
        className="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setOpen(false)}
        >
          X
        </button>
        {
          title
          && (
            <div className="w-full pb-4 border-b border-gunmetal-600">
              <span className="text-lg">
                {title}
              </span>
            </div>
          )
        }
        {children}
      </div>
    </div>
  );
}
