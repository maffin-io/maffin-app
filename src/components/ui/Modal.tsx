import React from 'react';
import RModal from 'react-modal';

export type ModalProps = {
  className?: string;
  triggerContent: string | JSX.Element;
  triggerClassName?: string;
  showClose?: boolean;
} & React.PropsWithChildren;

export interface ModalRef {
  openModal: () => void;
  closeModal: () => void;
}

function Modal({
  className = 'modal',
  triggerClassName = '',
  triggerContent,
  showClose = false,
  children,
}: ModalProps, ref: React.Ref<ModalRef>): JSX.Element {
  const [isOpen, setIsOpen] = React.useState<boolean>(false);

  // Expose actions to the parent
  React.useImperativeHandle(ref, () => ({
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  }));

  return (
    <>
      <RModal
        isOpen={isOpen}
        overlayClassName="overlay"
        className={className}
        onRequestClose={() => setIsOpen(false)}
      >
        {
          showClose
          && (
            <button
              type="button"
              className="float-right"
              onClick={() => setIsOpen(false)}
            >
              X
            </button>
          )
        }
        {children}
      </RModal>

      <button
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        className={triggerClassName}
      >
        {triggerContent}
      </button>
    </>
  );
}

export default React.forwardRef(Modal);
