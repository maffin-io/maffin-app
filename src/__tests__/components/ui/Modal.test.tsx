import React from 'react';
import RModal from 'react-modal';
import { fireEvent, render, screen } from '@testing-library/react';

import Modal from '@/components/ui/Modal';

jest.mock('react-modal', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="RModal">
      {props.children}
    </div>
  ),
));

describe('Modal', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders hidden modal on mount with trigger', async () => {
    const { container } = render(
      <Modal
        triggerContent={<span>open</span>}
      >
        <TestComponent />
      </Modal>,
    );

    expect(RModal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );

    expect(container).toMatchSnapshot();
  });

  it('opens modal when clicking the trigger', async () => {
    render(
      <Modal
        triggerContent={<span>open</span>}
      >
        <TestComponent />
      </Modal>,
    );

    const button = await screen.findByText('open');
    fireEvent.click(button);

    await screen.findByText('submit form');
    const modal = await screen.findByTestId('RModal');
    expect(RModal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );
    expect(modal).toMatchSnapshot();
  });

  it('closes modal when clicking the X button', async () => {
    render(
      <Modal
        triggerContent={<span>open</span>}
        showClose
      >
        <TestComponent />
      </Modal>,
    );

    const button = await screen.findByText('open');
    fireEvent.click(button);

    await screen.findByText('submit form');
    await screen.findByTestId('RModal');
    expect(RModal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: true,
      }),
      {},
    );

    const xButton = screen.getByRole('button', { name: 'X' });
    fireEvent.click(xButton);

    expect(RModal).toHaveBeenLastCalledWith(
      expect.objectContaining({
        isOpen: false,
      }),
      {},
    );
  });
});

function TestComponent({
  onSave = () => {},
}: {
  onSave?: Function,
}): JSX.Element {
  return <button onClick={() => onSave()} type="button">submit form</button>;
}
