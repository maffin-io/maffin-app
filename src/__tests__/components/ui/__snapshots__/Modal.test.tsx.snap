// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal opens modal when clicking the trigger 1`] = `
<div
  data-testid="RModal"
>
  <button
    type="button"
  >
    submit form
  </button>
</div>
`;

exports[`Modal renders hidden modal on mount with trigger 1`] = `
<div>
  <div
    data-testid="RModal"
  >
    <button
      type="button"
    >
      submit form
    </button>
  </div>
  <button
    type="button"
  >
    <span>
      open
    </span>
  </button>
</div>
`;
